File: single_list.c
#include<stdio.h>
#include<stdlib.h>

/* the code  is 
completed with operations like traversing the list
 creating list , insertion operation and a display 
function,deleting nodes and reversing the list*/
void create_node();
void display_node();
void insertnode();
void insertnode_begining();
void insertnode_end();
void insertnode_location();
void deletenode();
void deletenode_begining();
void deletenode_end();
void deletenode_location();
void reverse_list();

struct node{
int data;
struct node *link;
};

struct node *head,*newnode,*temp;
// temp is used to traverse through the list 
int main()
{
int choice;
printf("\n Linked list program.");
do{
printf("\n 1.Create node \n2.Delete node \n3.Insert node \n4.Display node \n5.Reverse list \n6.Quit \nEnter the choice:  ");
scanf("%d",&choice);
switch(choice){
case 1:create_node();break;
case 2:deletenode();break;
case 3:insertnode();break;
case 4:display_node();break;
case 5:reverse_list();break;
case 6:printf("\n Bye See you later :)");break;
default : printf("\n Enter something valid ");
}}while(choice!=6);
printf("\n");
return 0;
}


void create_node(){
head = 0;
int choice = 1;
while(choice){
newnode = (struct node *)malloc(sizeof(struct node));
printf("\n Enter the number to be added to the list:");
scanf("%d",&newnode->data);
newnode->link = 0;
if(head == 0){
head = newnode;
temp=newnode;
}
else{
temp ->link = newnode;
temp = newnode;}
printf("\n Do you like to continue ?<0 for no , 1 for yes>:");
scanf("%d",&choice);
}
}

void display_node(){
int count=0;
printf("\n Displaying the elements");
temp = head;
while(temp !=0){
printf("\n Data is %d\n", temp->data);
temp = temp->link;
count++;}
printf("\n Number of nodes is %d\n",count);
}

void insertnode(){
int choice;
printf("\n1. insert begin \n2.insert end \n 3. insert location \n choice:");
scanf("%d",&choice);
switch(choice){
case 1:
insertnode_begining();
break;
case 2:
insertnode_end();
break;
case 3:
insertnode_location();
break;

}}


void insertnode_begining(){
printf("\nInsertion operation in linked list");
newnode = (struct node*)malloc(sizeof(struct node));
printf("\n Enter the data");
scanf("%d",&newnode->data);
newnode->link = head;
head= newnode;
}

void insertnode_end(){
printf("\nInsertion operation in linked list");
newnode = (struct node*)malloc(sizeof(struct node));
printf("\n Enter the data");
scanf("%d",&newnode->data);
temp = head;
while(temp->link != 0)
{
temp = temp -> link;
 }
temp->link = newnode;
newnode->link = 0;
}

void insertnode_location(){
int pos;
int count = 0;
printf("\nInsertion operation in linked list");
printf("\nEnter the position where after  you want to enter the element");
scanf("%d",&pos);
temp = head;
 while (temp !=0){
temp = temp ->link;
count++;}
 if (pos> count){
printf("\n position invalid");}
else if (pos==1){
insertnode_begining();}
else{
int i=1;
temp = head;
newnode = (struct node*)malloc(sizeof(struct node));
printf("\nEnter the element to be inserted :");
scanf("%d",&newnode->data);
while(i<pos)
{
temp = temp -> link;
i++;}
newnode->link = temp -> link;
temp->link =  newnode;
}
}

void deletenode(){
int choice;
printf("\n1.Delete begining\n2.Delete at end\n3.Delete at position\nchoice:");
scanf("%d",&choice);
switch(choice){
case 1:
deletenode_begining();
break;
case 2:
deletenode_end();
break;
case 3:
deletenode_location();
break;
}}

void deletenode_begining(){
if (head == 0)
{
printf("\nList is empty");
}else{
temp = head;
head = head ->link;
free(temp);
printf("\nNode deleted");}
}

void deletenode_end(){
struct node *prev;
if (head == 0)
{
printf("\nList is empty");
}
else{
temp = head;
while(temp ->link != 0){
prev = temp;
temp = temp ->link;}
prev-> link = 0;
free(temp);
printf("\nNode deleted");
}
}

void deletenode_location(){
int count=0;
int pos;
int i=1;
struct node *next;
temp = head;
if (head == 0)
{
printf("\nList is empty");
}
while(temp !=0){
temp = temp ->link;
count++;}
printf("Enter the position you want to  enter the node:");
scanf("%d",&pos);
if(pos>count){
printf("\n Position innvalid");}
else if(pos==1){
deletenode_begining();}
else
{
temp = head;
while(i<pos-1)
{
temp = temp ->link;
i++;
}
next =  temp ->link;
temp ->link =  next ->link;
free(next);
printf("\nOne elemenet deleted");
}
}

void reverse_list(){
struct node   *prev,*current,*next;
prev = 0;
current = head;
next = head;
while(next!=0){
next = current ->link;
current->link = prev;
prev = current;
current = next;
}
head = prev;
}

File: SingleList.java
/* It is a fairly simple and understandable java code 
for performing some basic linked list operation like 
insertion, deletion, reversing and creation of list
TODO making the linked list to store value of different type 
*/


import java.util.*;

class SingleList{

Scanner sc = new Scanner(System.in);

Node head;
	
	class Node{
	int data;
	Node next;
		Node(int data)
		{
		this.data = data;
	}
}



void createNode(SingleList l){
System.out.print("Enter an integer data:");
int data = sc.nextInt();
Node newnode = new Node(data);
newnode.next = null;
if(l.head == null)
	l.head = newnode;
else{
	Node lastnode = l.head;
	while(lastnode.next != null)
	lastnode = lastnode.next;
	lastnode.next = newnode;	
}
}

void printList(SingleList l){
Node lastnode = l.head;
System.out.print("Linked List:");
while(lastnode != null){
	System.out.print(lastnode.data+" ");
	lastnode = lastnode.next;
}
System.out.print("\n");
}



void deleteNodeByValue(SingleList l){
System.out.print("Enter a value to delete:");
int val = sc.nextInt();
Node curren_node = l.head,prev =  null;
while(curren_node !=null && curren_node.data == val && curren_node==l.head){
	l.head = curren_node.next;
	curren_node =  curren_node.next;
	prev = curren_node;	
	System.out.println("Element deleted");
	}
while(curren_node !=null ){
	if(curren_node.data == val)
	prev.next = curren_node.next;
	else
	prev = curren_node;
	curren_node =  curren_node.next;
}
}


static public int sizeOfList(SingleList l){
Node curren_node =  l.head;
int count = 1;
while(curren_node != null){
	curren_node = curren_node.next;
	count++;
	}
return count;
}

static public boolean isEmpty(SingleList l){
return (sizeOfList(l) == 0) ? true : false ;
}

void deleteNodeByPosition(SingleList l){
System.out.print("Enter a position to perform deletion:");
int pos = sc.nextInt();
int count = 1;
Node curren_node = l.head,prev = null;
if(pos > sizeOfList(l)){ System.out.println("Element not found");}
else{
if(pos == 1){
	l.head = curren_node.next;
	curren_node.next = null;
	System.out.println("Element deleted on head");
}else{
while(pos != count){
	prev = curren_node;
	curren_node = curren_node.next;
	count +=1;
}	
if(pos == count){
	prev.next = curren_node.next;
	curren_node.next = null;
	System.out.println("Element deleted");
}
}
}
}


void insertNode(SingleList l){
System.out.print("Enter the position to insert data:");
int pos = sc.nextInt();
if((pos-1) > sizeOfList(l)){System.out.println("Position invalid");}
else{
System.out.print("Enter the data to be entered:");
int val = sc.nextInt();
Node new_node = new Node(val);
new_node.next = null;
Node curren_node = l.head;
int count = 1;
if(pos == 1){
	new_node.next = l.head;
	l.head = new_node;
	}
else{
	while(count != pos -1){
	curren_node = curren_node.next;
	count++;
	}	
	new_node.next = curren_node.next;
	curren_node.next = new_node;	
	}
}
}


void reverseList(SingleList l){
Node curren_node = l.head;
Node next_node = l.head;
Node prev_node =  null;
while(curren_node != null){
next_node = curren_node.next;
curren_node.next = prev_node;
prev_node =  curren_node;
curren_node = next_node;
}
l.head = prev_node;
System.out.println("List Reversed");
}

public static void main(String args[])
{
SingleList l = new SingleList();
Scanner sc = new Scanner(System.in);
int choice;
do{
System.out.print("--------------------------------\n");
System.out.print("1.Create node \n2.Delete node{Value} \n3.Delete node {Position}   \n4.Insert node \n5.ReverseList \n6.PrintList \n7.Quit \nEnter the choice:");
choice = sc.nextInt();
switch(choice){
case 1:l.createNode(l);break;
case 2:l.deleteNodeByValue(l);break;
case 3:l.deleteNodeByPosition(l);break;
case 4:l.insertNode(l);break;
case 5:l.reverseList(l);break;
case 6:l.printList(l);break;
case 7:System.out.print("Bye See you later :)\n");break;
default :System.out.print("Enter something valid\n");
}}while(choice != 7);
} 
}


